version-catalog/api/src/main/kotlin/io/spine/internal/catalog/entry/VersionInheritingEntry.kt

```diff
@@ -0,0 +1,142 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.internal.catalog.entry
+
+import io.spine.internal.catalog.Alias
+import io.spine.internal.catalog.CatalogRecord
+import io.spine.internal.catalog.VersionNotation
+import io.spine.internal.catalog.VersionRecord
+
+/**
+ * A version entry, which can take the version from the outer entry.
+ *
+ * From the outside it looks like the entry just "inherits" the version from
+ * its parent.
+ *
+ * This entry is not meant to be used directly, rather server as a foundation
+ * for other entries.
+ *
+ * In order to understand why it's needed, consider the following snippet:
+ *
+ * ```
+ * internal object Kotlin : LibraryEntry() {
+ *     override val version = "1.6.21"          // libs.versions.kotlin
+ *     override val module = "jetbrains:kotlin" // libs.kotlin
+ *
+ *     object Runtime : LibraryEntry() {
+ *         override val version = Kotlin.version            // libs.versions.kotlin.runtime
+ *         override val module = "jetbrains:kotlin-runtime" // libs.kotlin.runtime
+ *     }
+ * }
+ * ```
+ *
+ * Both `Kotlin` and `Kotlin.Runtime` libraries will receive the same version,
+ * declared in a single place. But for each library, an independent version
+ * record will be created:
+ *
+ *  1. `libs.versions.kotlin`.
+ *  2. `libs.versions.kotlin.runtime`.
+ *
+ * Thus, a local overriding (in settings file) of `libs.versions.kotlin` will
+ * not affect the version of `Kotlin.Runtime` library. But, intuitively should.
+ *
+ * In contrast, when using this entry, there's no need in manual declaration of
+ * the version in the nested entry. A version inheriting entry can take the version
+ * from the outer entry on its own.
+ *
+ * Consider the same snippet, but with an entry which extends this class:
+ *
+ * ```
+ * internal object Kotlin : LibraryEntry() {
+ *     override val version = "1.6.21"              // libs.versions.kotlin
+ *     override val module = "org.jetbrains:kotlin" // libs.kotlin
+ *
+ *     object Runtime : SomeVersionInheritingEntry() {
+ *         override val module = "org.jetbrains:kotlin-runtime" // libs.kotlin.runtime
+ *     }
+ * }
+ * ```
+ *
+ * Going this way, only `libs.versions.kotlin` will be generated and available
+ * for a local override. Both `Kotlin` and `Kotlin.Runtime` libraries will
+ * reference `libs.versions.kotlin` version. When this version is overridden,
+ * both libraries will be affected as well.
+ *
+ * Although, such entries are not bound to use only an inherited version.
+ * It is still possible to declare the version within this entry. But in case,
+ * when neither this entry nor its outer one declares a version, an exception
+ * will be thrown.
+ *
+ * See direct implementations: [PluginEntry], [DependencyEntry].
+ */
+open class VersionInheritingEntry : CatalogEntry(), VersionNotation {
```

Comment:
Let's discuss this vocally. From our last round, you said `CatalogEntry` is a counter-example for `VersionInheritingEntry`, as `CatalogEntry` takes the properties from its children, while `VersionInheritingEntry` descendants — at the moment — just take their versions from their parent. 

Two points:

1. We might need to inherit more than just a version. Hence the name is still questionable.
2. `CatalogEntry` is not a counter-example, but a direct parent type. Which makes your previous comments confusing to me.